### PR TITLE
Loom Launcher

### PR DESCRIPTION
**Pull request by Newtsuck**
**Some Preemptive Notes**

- It seems that somewhere along the line a bunch of files got deleted somehow
- The actual System.cpp lives in the common directory, and should probably find a better home
- Currently the class is simply called System, but perhaps a better name and namespace could be thought up?
- Within this are some workaround hacks for Text Inputs that require testing (for if they are still required)

### CODE REVIEW

File: application/common/System.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+#include "loom/common/utils/utString.h"
+#include "loom/common/platform//platformThread.h"
+#include <iostream>
+#include <stdio.h>
+#include <stdlib.h>
+
+class System
+{
+private:
+	static void zeroBuffer(char *buff)
```
Comment (bengarney) [inline application/common/System.cpp]: Is this the same as memset?

File: application/common/System.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
+#include "loom/script/loomscript.h"
+#include "loom/common/utils/utString.h"
+#include "loom/common/platform//platformThread.h"
+#include <iostream>
+#include <stdio.h>
+#include <stdlib.h>
+
+class System
```
Comment (bengarney) [inline application/common/System.cpp]: I think a class name like Process would be a lot more accurate than System. What do you think? @SmilyOrg @LukeLamothe thoughts?

File: application/common/System.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+		while (true/*!feof(self->pipe)*/) 
+		{
+			// Get the next character!
+			int recievedChar = fgetc(self->pipe);
+
+			// If getting the last character set the EOF flag, break out of the loop immedietly
+			if (feof(self->pipe)) break;
+
+			// Make sure we are getting actual characters
+			if (recievedChar < 0) continue;
```
Comment (bengarney) [inline application/common/System.cpp]: receive not recieve.

File: application/common/System.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	{
+		System* self = (System*)param;
+
+		// Zero out the buffer
+		zeroBuffer(self->buffer);
+
+		while (true/*!feof(self->pipe)*/) 
+		{
+			// Get the next character!
+			int recievedChar = fgetc(self->pipe);
```
Comment (bengarney) [inline application/common/System.cpp]: Should this just be unsigned char?

File: application/common/System.cpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+		return id;
+	}
+};
+
+
+static int registerLoomSystem(lua_State* L)
+{
+	beginPackage(L, "loom")
+
+		.beginClass<System>("natSystem")
```
Comment (bengarney) [inline application/common/System.cpp]: Shouldn't classes have capital names?

File: sdk/src/feathers/controls/TextInput.ls [hunk 6]
```diff
@@ -133,6 +133,12 @@ package feathers.controls
          * @private
          */
         protected var _text:String = "";
+        
+        /**
+         * @private
+         * Par of a Workaround fix that specially prevents editing when not in focus
```
Comment (bengarney) [inline sdk/src/feathers/controls/TextInput.ls]: Part?

Comment (bengarney) [inline sdk/src/feathers/controls/TextInput.ls]: @SmilyOrg Can you review this workaround and see if it should be rolled into master?

File: sdk/src/loom/System.ls [hunk 7]
```diff
@@ -0,0 +1,131 @@
+package loom
+{
+    import system.platform.Platform;
+    import system.platform.PlatformType;
+    
+    public class System
```
Comment (bengarney) [inline sdk/src/loom/System.ls]: Please fix the typos in this class. The documentation is fairly good otherwise.

Comment (bengarney) [inline sdk/src/loom/System.ls]: Does this need to wrap a natSystem or can it just BE a natSystem with extra LoomScript goodies like String?

File: sdk/src/loom/System.ls [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        
+        /**
+         * Constructor
+         * 
+         * The constructor takes an optional ID, that can be used to distinguish multiple system objects
+         * using the special cmdWithId delegate.
+         * 
+         * @param id The Id to be given to this system object. Defaults to -1.
+         */
+        public function System(id:int = -1)
```
Comment (bengarney) [inline sdk/src/loom/System.ls]: Should the ID be implicitly generated?

File: sdk/src/loom/System.ls [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+         * delegate. Data is automatically extracted from both STDOUT and STDERR.
+         * 
+         * NOTE: If you wish to run a synchronous multi-part command, it is reccomended that you use the multCmd function because of
+         * platform specific differences in command syntax.
+         * 
+         * @param command The command that will be run through the native command interface.
+         * 
+         * @see multCmd
+         */
+        public function cmd(command:String):void { _system.cmd(command + " 2>&1"); }
```
Comment (bengarney) [inline sdk/src/loom/System.ls]: Is the 2>&1 idiom portable?

Should this be called run or something? Look at the Adobe AIR API for this.

File: sdk/src/loom/System.ls [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        
+        /**
+         * Synchronously runs the provided commands in order and gets feedback from that command. Functionally simmilar to cmd, but 
+         * this function takes into account platform specific syntax differences to run multiple commands. Data is automatically extracted
+         * from both STDOUT and STDERR.
+         * 
+         * @param commands A vector of strings that contains all of the commands to be synchronously run. The commands in this vector will
+         * be run in order, starting at index 0.
+         */
+        public function multCmd(commands:Vector.<String>):void
```
Comment (bengarney) [inline sdk/src/loom/System.ls]: If a command fails are the following commands run?

File: sdk/src/loom/System.ls [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        * new line character, or was over 1024 characters long. In the case that the line was simply too long
+        * the string will be broken up and returned over multiple calls of this delegate.
+        *
+        * NOTE: While a chunk of data is defined by end of line characters. The returned data will NOT contain
+        * end of line characters.
+        * 
+        * @param val The function to be added to the onData delegate. It is expected that this function will have
+        * the signature `function(String):void`.
+        */
+        public function addOnData(val:Function):void { _system.onData += val; }
```
Comment (bengarney) [inline sdk/src/loom/System.ls]: Why is this needed? (Same for next 3 methods.)

File: application/common/System.cpp [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+public:
+
+	LOOM_DELEGATE(OnData);
+	LOOM_DELEGATE(OnFinish);
+
+	System(int i)
+	{
+		// Set the ID at construction
+		id = i;
```
Comment (bengarney) [inline application/common/System.cpp]: Why is the ID needed, given that users can provide any data they like to the delegate listener?

**Comment by bengarney** [comment]
If you do 

git checkout master docs/vendor/gems

it should restore the mysteriously deleted files.

File: sdk/src/loom/System.ls (hunk 7 above)
Comment (Newtsuck) [inline sdk/src/loom/System.ls]: The issue I ran into is with the inclusion of the delegates. It seems that because the class has native delegates, the class itself must be native, and native classes can't have any extra goodies inside of them.

File: sdk/src/loom/System.ls (hunk 9 above)
Comment (Newtsuck) [inline sdk/src/loom/System.ls]: 1. Yes, the 2>&1 works the same for Mac and Windows
2. That sounds good

File: sdk/src/loom/System.ls (hunk 10 above)
Comment (Newtsuck) [inline sdk/src/loom/System.ls]: It appears that on windows they are, but on Mac they are not

File: application/common/System.cpp (hunk 4 above)
Comment (Newtsuck) [inline application/common/System.cpp]: The fgetc function actually returns an integer type ([According to the C++ documentation](http://www.cplusplus.com/reference/cstdio/fgetc/)), so I keep it as an integer since I am just using the integer representations of the line ending characters anyway

File: sdk/src/loom/System.ls (hunk 11 above)
Comment (Newtsuck) [inline sdk/src/loom/System.ls]: Has to do with the native delegates. It seems that in order to include a native delegate the class itself must be native which means there can be no additions to code on the loom side. I also like this implementation because it allows for user supplied delegate functions to simply not have an ID parameter if the user doesn't care about the ID.

File: application/common/System.cpp (hunk 12 above)
Comment (Newtsuck) [inline application/common/System.cpp]: The ID is meant to assist in determining from what process data is being received when data is being received asynchronously from multiple process objects. If there is another way to determine that then I would very much like to learn it!

File: application/common/System.cpp (hunk 4 above)
Comment (bengarney) [inline application/common/System.cpp]: Hmm. Yes - if you want to check for EOF return value you must keep it an int. But you don't do that - you only compare to values in unsigned char range. It doesn't really matter - you can leave it as is.

File: application/common/System.cpp (hunk 12 above)
Comment (bengarney) [inline application/common/System.cpp]: There is another way.

```
var id1:int = 100;
myObject.onEvent += function():void { trace(id1); };
var id2:int = 200;
myObject.onEvent += function():void { trace(id2); };
```

The first function will hold a reference to id1 and print 100. The second will print 200 due to referencing id2. You could also reference myObject instead. Check out closurize() for other cool tricks.

Comment (bengarney) [inline application/common/System.cpp]: Note this is preferable because it lets you reference arbitrary and unbounded items with no additional complexity in the supporting C++ code.

File: sdk/src/loom/System.ls (hunk 7 above)
Comment (bengarney) [inline sdk/src/loom/System.ls]: What about DisplayObject? https://github.com/LoomSDK/LoomSDK/blob/master/sdk/src/loom2d/Display/DisplayObject.ls#L214

File: sdk/src/loom/System.ls (hunk 10 above)
Comment (bengarney) [inline sdk/src/loom/System.ls]: We should probably normalize that if we can. It's a key implementation detail. At the worst we have to document the difference in behavior.

File: sdk/src/loom/System.ls (hunk 11 above)
Comment (bengarney) [inline sdk/src/loom/System.ls]: Hmm - how does it allow the parameter to be treated differently than the delegate normally would?

**Comment by bengarney** [comment]
Great. Thank you! I responded to your comments.

File: application/common/System.cpp (hunk 2 above)
Comment (Newtsuck) [inline application/common/System.cpp]: There is already a class called [Process](https://github.com/LoomSDK/LoomSDK/blob/dfe41a7b508b23c4695a40023ebd523b1c4ea655/sdk/src/system/Process.ls), perhaps this functionality should be added to that class?

Comment (SmilyOrg) [inline application/common/System.cpp]: I think adding it to Process would be prudent. Maybe we can model it after the [AIR API](http://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/desktop/NativeProcess.html).

File: application/common/System.cpp (hunk 1 above)
Comment (SmilyOrg) [inline application/common/System.cpp]: Probably `memset(buff, 0, strlen(buff))`. But if you just want to empty a string, `buff[0] = 0` should work too?

File: sdk/src/feathers/controls/TextInput.ls (hunk 6 above)
Comment (SmilyOrg) [inline sdk/src/feathers/controls/TextInput.ls]: I fixed a bunch of text input / focus bugs most likely after this was integrated, so I would say that it would be good to temporarily revert this piece of code, merge in master (if not already) and check if there are still any issues with input.
